addons/sale_mrp/models/sale_mrp.py

```diff
@@ -10,20 +10,51 @@ class SaleOrderLine(models.Model):
-                    if bom_delivered:
-                        line.qty_delivered = line.product_uom_qty
+        for order_line in self:
+            if order_line.qty_delivered_method == 'stock_move':
+                boms = order_line.move_ids.mapped('bom_line_id.bom_id')
+                # We fetch the BoMs of type kits linked to the order_line,
+                # the we keep only the ones related to the finished produst.
+                relevant_boms = boms.filtered(lambda b: b.type == 'phantom' and (b.product_id == order_line.product_id or b.product_tmpl_id == order_line.product_id.product_tmpl_id))
+                if relevant_boms:
+                    # In case  multiples BoMs of type kits are set on the finished product
+                    prioritary_bom = relevant_boms[0]
+                    moves = order_line.move_ids.filtered(lambda m: m.state == 'done' and not m.scrapped)
+                    qty_ratios = []
+                    order_uom_qty = order_line.product_uom._compute_quantity(order_line.product_uom_qty, prioritary_bom.product_uom_id)
+                    boms, bom_sub_lines = prioritary_bom.explode(order_line.product_id, order_uom_qty)
+                    for bom_line, bom_line_data in bom_sub_lines:
+                        relevant_moves = moves.filtered(lambda m: m.bom_line_id == bom_line)
+                        if relevant_moves:
+                            # We compute the quantities needed of each components to make one kit.
+                            # Then, we collect every relevant moves related to a specific component
+                            # to know how many are considered delivered.
+                            qty_needed = bom_line_data['qty'] / bom_line_data['original_qty']
+                            qty_uom_needed = bom_line.product_uom_id._compute_quantity(qty_needed, bom_line.product_id.uom_id)
+                            qty_uom_processed = 0.0
+                            relevant_delivered_moves = relevant_moves.filtered(lambda m: m.location_dest_id.usage == 'customer' and not m.origin_returned_move_id or not (m.origin_returned_move_id and m.to_refund))
+                            for move in relevant_delivered_moves:
+                                qty_uom_processed += move.product_uom._compute_quantity(move.quantity_done, move.product_tmpl_id.uom_id)
+                            relevant_return_moves = relevant_moves.filtered(lambda m: m.location_dest_id.usage != 'customer' and m.to_refund)
+                            for move in relevant_return_moves:
+                                qty_uom_processed -= move.product_uom._compute_quantity(move.quantity_done, move.product_tmpl_id.uom_id)
+                            # We compute a ratio to know how many kits we can produce with this quantity of that specific component
+                            qty_ratios.append(qty_uom_processed / qty_uom_needed)
+                        else:
```

Comment:
You could do a perf improvement here. If one of the element is 0 the min(qty_ratios) will be 0.